lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,291 @@
+		);
+	}
+
+	/**
+	 * Retrieves the document title from a remote URL.
+	 *
+	 * @param string $url The website url whose HTML we want to access.
+	 * @return string|WP_Error The URL's document title on success, WP_Error on failure.
+	 */
+	private function get_remote_url( $url ) {
+
+		$args = array(
+			'limit_response_size' => 150 * KB_IN_BYTES,
+		);
+
+		/**
+		 * Filters the HTTP request args for URL data retrieval.
+		 *
+		 * Can be used to adjust response size limit and other WP_Http::request args.
+		 *
+		 * @param array $args Arguments used for the HTTP request
+		 * @param string $url The attempted URL.
+		 */
+		$args = apply_filters( 'rest_url_details_http_request_args', $args, $url );
+
+		$response = wp_safe_remote_get(
+			$url,
+			$args
+		);
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'no_response', __( 'URL not found. Response returned a non-200 status code for this URL.', 'gutenberg' ), array( 'status' => WP_Http::NOT_FOUND ) );
+		}
+
+		$remote_body = wp_remote_retrieve_body( $response );
+
+		if ( empty( $remote_body ) ) {
+			return new WP_Error( 'no_content', __( 'Unable to retrieve body from response at this URL.', 'gutenberg' ), array( 'status' => WP_Http::NOT_FOUND ) );
+		}
+
+		return $remote_body;
+	}
+
+	/**
+	 * Parses the <title> contents from the provided HTML
+	 *
+	 * @param string $html the HTML from the remote website at URL.
+	 * @return string the title tag contents (maybe empty).
+	 */
+	private function get_title( $html ) {
+		preg_match( '|<title>([^<]*?)</title>|is', $html, $match_title );
+
+		$title = isset( $match_title[1] ) ? trim( $match_title[1] ) : '';
+
+		return $title;
+	}
+
+	/**
+	 * Utility function to build cache key for a given URL.
+	 *
+	 * @param string $url the URL for which to build a cache key.
+	 * @return string the cache key.
+	 */
+	private function build_cache_key_for_url( $url ) {
+		return 'g_url_details_response_' . md5( $url );
+	}
+
+	/**
+	 * Utility function to retrieve a value from the cache at a given key.
+	 *
+	 * @param string $key the cache key.
+	 * @return string the value from the cache.
+	 */
+	private function get_cache( $key ) {
+		return get_transient( $key );
+	}
+
+	/**
+	 * Utility function to cache a given data set at a given cache key.
+	 *
+	 * @param string $key the cache key under which to store the value.
+	 * @param array  $data the data to be stored at the given cache key.
+	 * @return void
+	 */
+	private function set_cache( $key, $data = array() ) {
+		if ( ! is_array( $data ) ) {
+			return;
+		}
+
+		/**
+		 * Filters the cache expiration.
+		 *
+		 * Can be used to adjust the time until expiration in seconds for the cache
+		 * of the data retrieved for the given URL.
+		 *
+		 * @param int HOUR_IN_SECONDS the time until cache expiration in seconds.
```

Comment:
```suggestion
		 * @param int $ttl The time until cache expiration in seconds.
```

This needs to be a variable name.

Reply:
Done